src/hardware/aim-tti-dps/api.c

```diff
@@ -10,132 +10,492 @@
+	sdi->inst_type = SR_INST_SCPI;
+	ret = sr_scpi_connection_id(scpi, &sdi->connection_id);
+	if (ret != SR_OK) {
+		g_free(sdi->connection_id);
+		sdi->connection_id = NULL;
+	}
+	sr_scpi_hw_info_free(hw_info);
+	hw_info = NULL;
+
+	for (chidx = 0; chidx < models[model_idx].channels; ++chidx) {
+		cg = g_malloc0(sizeof(struct sr_channel_group));
+		snprintf(buf, ARRAY_SIZE(buf), "CH%d", chidx + 1);
+		cg->name = g_strdup(buf);
+		sdi->channel_groups = g_slist_append(sdi->channel_groups, cg);
+
+		snprintf(buf, ARRAY_SIZE(buf), "V%d", chidx + 1);
+		ch = sr_channel_new(sdi, 2 * chidx, SR_CHANNEL_ANALOG, TRUE, buf);
+		cg->channels = g_slist_append(cg->channels, ch);
+
+		snprintf(buf, ARRAY_SIZE(buf), "I%d", chidx + 1);
+		ch = sr_channel_new(sdi, 2 * chidx + 1, SR_CHANNEL_ANALOG, TRUE, buf);
+		cg->channels = g_slist_append(cg->channels, ch);
+		cg->priv = malloc(sizeof(int));
+		*(int*)(cg->priv) = chidx;
+	}
 
-	return devices;
+	devc = g_malloc0(sizeof(struct dev_context));
+	sr_sw_limits_init(&(devc->limits));
+	devc->model_config = &models[model_idx];
+	devc->config = g_malloc0(sizeof(struct per_channel_dev_context) *
+									models[model_idx].channels);
+	sdi->priv = devc;
+
+	if (aim_tti_dps_sync_state(scpi, devc) < 0) {
+		sr_scpi_close(sdi->conn);
+		sr_dev_inst_free(sdi);
+		g_free(devc);
+		sr_dbg("Scan failed.");
+		return NULL;
+	}
+	return sdi;
+}
+
+static GSList *scan(struct sr_dev_driver *di, GSList *options)
+{
+	return sr_scpi_scan(di->context, options, probe_device);
 }
 
 static int dev_open(struct sr_dev_inst *sdi)
 {
-	(void)sdi;
+	struct sr_scpi_dev_inst *scpi;
+	int ret;
+
+	if (!sdi || !sdi->conn)
+		return SR_ERR_BUG;
+	scpi = sdi->conn;
 
-	/* TODO: get handle from sdi->conn and open it. */
+	ret = sr_scpi_open(scpi);
+	if (ret < 0) {
+		sr_err("Failed to open SCPI device: %s.", sr_strerror(ret));
+		return SR_ERR;
+	}
 
 	return SR_OK;
 }
 
 static int dev_close(struct sr_dev_inst *sdi)
 {
-	(void)sdi;
+	struct sr_scpi_dev_inst *scpi;
 
-	/* TODO: get handle from sdi->conn and close it. */
+	if (!sdi || !sdi->conn)
+		return SR_ERR_BUG;
+	scpi = sdi->conn;
 
-	return SR_OK;
+	if (sdi->status <= SR_ST_INACTIVE)
+		return SR_OK;
+
+	/* enable local interface again */
+	sr_scpi_send(scpi, "LOCAL");
+
+	return sr_scpi_close(sdi->conn);
 }
 
 static int config_get(uint32_t key, GVariant **data,
 	const struct sr_dev_inst *sdi, const struct sr_channel_group *cg)
 {
-	int ret;
-
-	(void)sdi;
-	(void)data;
-	(void)cg;
-
-	ret = SR_OK;
-	switch (key) {
-	/* TODO */
-	default:
-		return SR_ERR_NA;
+	struct dev_context *devc;
+	int channel;
+
+	if (!sdi || !sdi->priv || !data)
+		return SR_ERR_ARG;
+
+	devc = sdi->priv;
+
+	if (!cg) {
+		switch (key) {
+		case SR_CONF_LIMIT_SAMPLES:
+		case SR_CONF_LIMIT_MSEC:
+			return sr_sw_limits_config_get(&devc->limits, key, data);
+		case SR_CONF_CONN:
+			*data = g_variant_new_string(sdi->connection_id);
+			break;
+		case SR_CONF_ENABLED:
+			gboolean any_output_enabled = FALSE;
```

Comment:
Please declare variables at the top of the function, and initialize later (or here).